lib/src/devtools_integration/_envelopes.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../model.dart';
+import 'messages.dart';
+
+/// Generic parameter is not used for encoder, because the message type cannot be detected in runtime.
+typedef AppMessageEncoder = Map<String, dynamic> Function(dynamic message);
+
+typedef AppMessageDecoder<T> = T Function(Map<String, dynamic> message);
+
+enum Channel {
+  requestToApp,
+  eventFromApp,
+  responseFromApp,
+}
+
+enum Codes {
+  // Events from app.
+  started,
+  summary,
+
+  // Requests to app.
+  detailsRequest,
+
+  // Successfull responses from app.
+  leakDetails,
+
+  // Error responses from app.
+  leakTrackingTurnedOffError,
+  unexpectedError,
+  unexpectedRequestTypeError,
+  ;
+
+  static Codes byName(String name) =>
+      Codes.values.where((e) => e.name == name).single;
+}
+
+class _JsonFields {
+  static const envelopeCode = 'code';
+  static const content = 'content';
+}
+
+class Envelope<T> {
```

Comment:
Some information about could be helpful here:

- what is an envelope?
- what is it used for?
- how do we expect it to behave?
- any other info that might be helpful